src/track/beatmap.cpp

```diff
@@ -457,11 +467,28 @@ double BeatMap::getBpmAroundPosition(double curSample, int n) const {
         }
     }
 
-    Beat startBeat, stopBeat;
-    startBeat.set_frame_position(
-            static_cast<google::protobuf::int32>(samplesToFrames(lower_bound)));
-    stopBeat.set_frame_position(static_cast<google::protobuf::int32>(samplesToFrames(upper_bound)));
-    return calculateBpm(startBeat, stopBeat);
+    VERIFY_OR_DEBUG_ASSERT(lower_bound < upper_bound) {
+        return -1;
+    }
+
+    const int kFrameEpsilon = m_iSampleRate / 20;
+
+    int numberOfBeats = 0;
+    for (const auto& beat : m_beats) {
+        double pos = framesToSamples(beat.frame_position() + kFrameEpsilon);
+        if (pos > upper_bound) {
+            break;
+        }
+        if (pos > lower_bound) {
+            numberOfBeats++;
+        }
+    }
+
+    if (numberOfBeats == 0) {
+        return -1;
+    }
+
+    return 60.0 * numberOfBeats * m_iSampleRate * kFrameSize / (upper_bound - lower_bound);
```

Comment:
It is hard to discover that this code actually duplicates a special case calculation in BeatUtils. Even the magic number 12 is copied literally here.

I don't recommend to introduce such a maintenance nightmare. Please add the missing code to BeatUtils and use it consistently from both locations.

Reply:
OK, I have used now a common function. I have removed asserting for 12 beats, because that is kind of pointless here. The usage code should be in charge to decide for a reasonable window. 

IMHO this code should be removed anyway, because a gradual speed changed sounds bad. It is better to allign speed changes with the a down beat. ... in future.